core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java

```diff
@@ -197,6 +215,290 @@ public Page getOutput()
+            this.taskFailureListener = requireNonNull(taskFailureListener, "taskFailureListener is null");
+            this.retryPolicy = requireNonNull(retryPolicy, "retryPolicy is null");
+            this.exchangeManagerRegistry = requireNonNull(exchangeManagerRegistry, "exchangeManagerRegistry is null");
+        }
+
+        @Override
+        public SerializedPage pollPage()
+        {
+            if (closed.get()) {
+                return null;
+            }
+
+            ExchangeDataSource dataSource = delegate.get();
+            if (dataSource == null) {
+                return null;
+            }
+            return dataSource.pollPage();
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            if (closed.get()) {
+                return true;
+            }
+            ExchangeDataSource dataSource = delegate.get();
+            if (dataSource == null) {
+                return false;
+            }
+            return dataSource.isFinished();
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            if (closed.get()) {
+                return immediateVoidFuture();
+            }
+            if (!initializationFuture.isDone()) {
+                return initializationFuture;
+            }
+            ExchangeDataSource dataSource = delegate.get();
+            if (dataSource == null) {
+                return immediateVoidFuture();
+            }
+            return dataSource.isBlocked();
+        }
+
+        @Override
+        public void addSplit(RemoteSplit remoteSplit)
+        {
+            SettableFuture<Void> future = null;
+            synchronized (this) {
+                if (closed.get()) {
+                    return;
+                }
+                ExchangeDataSource dataSource = delegate.get();
+                if (dataSource == null) {
+                    if (remoteSplit.getExchangeInput() instanceof DirectExchangeInput) {
+                        DirectExchangeClient directExchangeClient = directExchangeClientSupplier.get(systemMemoryContext, taskFailureListener, retryPolicy);
+                        dataSource = new DirectExchangeDataSource(directExchangeClient);
+                        dataSource.addSplit(remoteSplit);
+                    }
+                    else if (remoteSplit.getExchangeInput() instanceof ExternalExchangeInput) {
+                        ExternalExchangeInput externalExchangeInput = (ExternalExchangeInput) remoteSplit.getExchangeInput();
+                        ExchangeManager exchangeManager = exchangeManagerRegistry.getExchangeManager();
+                        ExchangeSource exchangeSource = exchangeManager.createSource(externalExchangeInput.getExchangeSourceHandles());
+                        dataSource = new ExternalExchangeDataSource(exchangeSource, systemMemoryContext);
+                    }
```

Comment:
It feels that we should be able to statically know (when OperatorFactory is created) what kind o splits we will be getting. Right?

Reply:
Yeah, in theory we could send the information about the implementation of an exchange as part of the plan node and than based on that information create an operator that knows how to consume an exchange of a specific type. 

However I'm not sure if we want to make this decision planning time or scheduling time? In theory it feels like the way how the data is exchange is an implementation detail that shouldn't be expressed in the plan.  In practice delaying this decision to the scheduling phase might give us more flexibility in the future (think of an adaptive execution where some stages are executed as streaming and some stages are executed with an external exchange).